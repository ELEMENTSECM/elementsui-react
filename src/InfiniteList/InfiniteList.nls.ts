export const nls = {
	en: {
		pullDownToRefresh: "Pull down to refresh",
		releaseToRefresh: "Release to refresh"
	},
	nb: {
		pullDownToRefresh: "Pull down to refresh",
		releaseToRefresh: "Release to refresh"
	},
	nn: {
		pullDownToRefresh: "Pull down to refresh",
		releaseToRefresh: "Release to refresh"
	},
	sv: {
		pullDownToRefresh: "Pull down to refresh",
		releaseToRefresh: "Release to refresh"
	}
};
