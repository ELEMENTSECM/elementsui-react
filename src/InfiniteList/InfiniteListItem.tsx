import * as React from "react";
import styled from "styled-components";

export interface InfiniteListItemProps {
	/** List item class name */
	className?: string;
	/** HTML element tag that will rendered for this item. Default: <li> */
	tag?: keyof JSX.IntrinsicElements;
	/** Any data that will be passed as event payload that could help identify the item */
	itemId?: any;
	/** onFocus event handler */
	onFocus?: (index?: number) => void;
	/** onClick event handler */
	onClick?: (index?: number) => void;
	/** onKeyDown event handler */
	onKeyDown?: (e: any, index?: number) => void;
	/** Tabindex*/
	tabIndex?: number;
}

const ListItem = styled.li`
	overflow: hidden;
	position: relative;
`;

const InfiniteListItem: React.SFC<InfiniteListItemProps> = ({
	children,
	className,
	tag,
	onFocus,
	onClick,
	onKeyDown,
	itemId,
	tabIndex = 0
}) => {
	const onFocusCallback = React.useCallback(() => onFocus?.(itemId), [onFocus, itemId]);
	const onClickCallback = React.useCallback(() => onClick?.(itemId), [onClick, itemId]);
	const onKeyDownCallback = React.useCallback((e) => onKeyDown?.(e, itemId), [onKeyDown, itemId]);
	return (
		<ListItem as={tag} tabIndex={tabIndex} className={className} role="listitem" onFocus={onFocusCallback} onClick={onClickCallback} onKeyDown={onKeyDownCallback}>
			{children}
		</ListItem>
	);
};

export default React.memo<React.PropsWithChildren<InfiniteListItemProps>>(InfiniteListItem);
