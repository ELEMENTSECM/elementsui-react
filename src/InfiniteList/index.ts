export { default, InfiniteListProps } from "./InfiniteList";
export { default as ListItem, InfiniteListItemProps } from "./InfiniteListItem";
