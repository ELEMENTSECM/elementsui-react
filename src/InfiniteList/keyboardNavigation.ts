import * as React from "react";
import { KeyMap } from "react-hotkeys";
import { find, isNil, isEmpty, takeWhile, partial, first, last } from "lodash";

export const keyMap: KeyMap = {
	down: "down",
	up: "up",
	left: "left",
	right: "right",
	pageDown: "pagedown",
	pageUp: "pageup",
	top: "home",
	bottom: "end"
} as const;

type NextItem = (start: HTMLElement, itemsPerRow: number) => HTMLElement | undefined;
type StepItem = (start: HTMLElement) => HTMLElement | undefined;

const currentItem = (list: HTMLElement | null) =>
	find(
		list?.children,
		x => x === document.activeElement || x.contains(document.activeElement)
	) as HTMLElement;
const itemsPerRow = (list: HTMLElement | null) => {
	if (isEmpty(list?.children)) {
		return 0;
	} else {
		const top = list?.children?.[0]?.getBoundingClientRect().top;
		return takeWhile(list?.children, child => child.getBoundingClientRect().top === top).length;
	}
};

const next = (step: StepItem, stepCount: number, start: HTMLElement) => {
	let remained = stepCount;
	let cursor = start;
	while (remained-- > 0) {
		const next = step(cursor);
		if (isNil(next)) {
			return cursor;
		} else if (remained === 0) {
			return next;
		} else {
			cursor = next!;
		}
	}
};

const stepForward: StepItem = x => x?.nextElementSibling as HTMLElement;
const stepBackward: StepItem = x => x?.previousElementSibling as HTMLElement;

export const handlers = (listRef: React.RefObject<HTMLElement>, { pageSize = 3 }) => {
	const navigate = (nextItem: NextItem) => (e?: KeyboardEvent) => {
		const item = nextItem(currentItem(listRef.current), itemsPerRow(listRef.current));
		if (item) {
			item.focus();
			e?.stopPropagation();
			e?.preventDefault();
		}
	};

	const left = navigate((start, itemsPerRow) => (itemsPerRow > 1 ? stepBackward(start) : undefined));
	const right = navigate((start, itemsPerRow) => (itemsPerRow > 1 ? stepForward(start) : undefined));
	const up = navigate((start, itemsPerRow) => next(stepBackward, itemsPerRow, start));
	const down = navigate((start, itemsPerRow) => next(stepForward, itemsPerRow, start));
	const pageDown = navigate(partial(next, stepForward, pageSize));
	const pageUp = navigate(partial(next, stepBackward, pageSize));
	const top = navigate(() => first(listRef.current?.children) as HTMLElement);
	const bottom = navigate(() => last(listRef.current?.children) as HTMLElement);

	return {
		left,
		right,
		up,
		down,
		pageDown,
		pageUp,
		top,
		bottom
	};
};
