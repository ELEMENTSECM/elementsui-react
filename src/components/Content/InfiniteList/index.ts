export { default, IInfiniteListProps } from "./InfiniteList";
