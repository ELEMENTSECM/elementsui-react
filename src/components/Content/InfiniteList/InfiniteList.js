import * as React from 'react';
import PropTypes from 'prop-types';
import { styles } from './InfiniteList.styles';
import { classNamesFunction, customizable, styled } from 'office-ui-fabric-react/lib/Utilities';
import Spinner from '../../Indicators/Spinner';
import { IntlProvider, FormattedMessage } from 'react-intl';
import nls from './InfiniteList.nls.json';

/** InfiniteList example */
class InfiniteListComponent extends React.Component {
	static getDerivedStateFromProps(props, state) {
		if (props.dataLength === state.dataLength) return null;

		return {
			dataLength: props.dataLength,
			showLoader: false,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};
	}

	constructor(props) {
		super();
		this.state = {
			dataLength: props.dataLength,
			showLoader: false,
			lastScrollTop: 0,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};

		this.startY = 0;
		this.currentY = 0;
		this.dragging = false;
		this.maxPullDownDistance = 0;

		this.throttledOnScrollListener = this.throttle(this.onScrollListener, 150);
		this.classNames = classNamesFunction()(styles, props);
	}

	componentDidMount() {
		this._scrollableNode = this.getScrollableTarget();
		this.el = this.props.height ? this._infScroll : this._scrollableNode || window;
		this.el.addEventListener('scroll', this.throttledOnScrollListener);

		if (
			typeof this.props.initialScrollY === 'number' &&
			this.el.scrollHeight > this.props.initialScrollY
		) {
			this.el.scrollTo(0, this.props.initialScrollY);
		}

		if (this.props.pullDownToRefresh) {
			this.el.addEventListener('touchstart', this.onStart);
			this.el.addEventListener('touchmove', this.onMove);
			this.el.addEventListener('touchend', this.onEnd);

			this.el.addEventListener('mousedown', this.onStart);
			this.el.addEventListener('mousemove', this.onMove);
			this.el.addEventListener('mouseup', this.onEnd);

			this.maxPullDownDistance = this._pullDown.firstChild.getBoundingClientRect().height;
			this.forceUpdate();

			if (typeof this.props.refreshFunction !== 'function') {
				throw new Error(
					`Mandatory prop "refreshFunction" missing.
					Pull Down To Refresh functionality will not work
					as expected.`
				);
			}
		}
	}

	componentWillUnmount() {
		this.el.removeEventListener('scroll', this.throttledOnScrollListener);

		if (this.props.pullDownToRefresh) {
			this.el.removeEventListener('touchstart', this.onStart);
			this.el.removeEventListener('touchmove', this.onMove);
			this.el.removeEventListener('touchend', this.onEnd);

			this.el.removeEventListener('mousedown', this.onStart);
			this.el.removeEventListener('mousemove', this.onMove);
			this.el.removeEventListener('mouseup', this.onEnd);
		}
	}

	getScrollableTarget = () => {
		if (this.props.scrollableTarget instanceof HTMLElement) return this.props.scrollableTarget;
		if (typeof this.props.scrollableTarget === 'string') {
			return document.getElementById(this.props.scrollableTarget);
		}
		if (this.props.scrollableTarget === null) {
			console.warn(`You are trying to pass scrollableTarget but it is null. This might
			happen because the element may not have been added to DOM yet.`);
		}
		return null;
	};

	onStart = evt => {
		if (this.state.lastScrollTop) return;

		this.dragging = true;
		this.startY = evt.pageY || evt.touches[0].pageY;
		this.currentY = this.startY;

		this._infScroll.style.willChange = 'transform';
		this._infScroll.style.transition = `transform 0.2s cubic-bezier(0,0,0.31,1)`;
	};

	onMove = evt => {
		if (!this.dragging) return;
		this.currentY = evt.pageY || evt.touches[0].pageY;

		if (this.currentY < this.startY) return;

		const threshold = this.props.pullDownToRefreshThreshold || 100;
		if (this.currentY - this.startY >= threshold) {
			this.setState(ps => ({
				...ps,
				pullToRefreshThresholdBreached: true
			}));
		}

		if (this.currentY - this.startY > this.maxPullDownDistance * 1.5) return;

		this._infScroll.style.overflow = 'visible';
		this._infScroll.style.transform = `translate3d(0px, ${this.currentY - this.startY}px, 0px)`;
	};

	onEnd = evt => {
		this.startY = 0;
		this.currentY = 0;

		this.dragging = false;

		if (this.state.pullToRefreshThresholdBreached) {
			this.props.refreshFunction && this.props.refreshFunction();
		}

		requestAnimationFrame(() => {
			if (this._infScroll) {
				this._infScroll.style.overflow = 'auto';
				this._infScroll.style.transform = 'none';
				this._infScroll.style.willChange = 'none';
			}
		});
	};

	isElementAtBottom(target, scrollThreshold = 0.8) {
		const clientHeight =
			target === document.body || target === document.documentElement
				? window.screen.availHeight
				: target.clientHeight;

		return target.scrollTop + clientHeight >= scrollThreshold * target.scrollHeight;
	}

	onScrollListener = event => {
		if (typeof this.props.onScroll === 'function') {
			setTimeout(() => this.props.onScroll(event), 0);
		}

		let target =
			this.props.height || this._scrollableNode
				? event.target
				: document.documentElement.scrollTop
					? document.documentElement
					: document.body;

		if (this.state.actionTriggered) return;

		let atBottom = this.isElementAtBottom(target, this.props.scrollThreshold);

		if (atBottom && this.props.hasMore) {
			this.setState(ps => ({ ...ps, actionTriggered: true, showLoader: true }));
			this.props.next();
		}
		this.setState(ps => ({ ...ps, lastScrollTop: target.scrollTop }));
	};

	throttle = (fn, threshhold, scope) => {
		threshhold || (threshhold = 250);
		let last, deferTimer;
		return function() {
			const context = scope || this;

			const now = +new Date(),
				args = arguments;
			if (last && now < last + threshhold) {
				clearTimeout(deferTimer);
				deferTimer = setTimeout(function() {
					last = now;
					fn.apply(context, args);
				}, threshhold);
			} else {
				last = now;
				fn.apply(context, args);
			}
		};
	};

	render() {
		const style = {
			height: this.props.height || 'auto'
		};
		const hasChildren = this.props.hasChildren || !!(this.props.children && this.props.children.length);

		const outerDivStyle = this.props.pullDownToRefresh && this.props.height ? { overflow: 'auto' } : {};
		const { root, pulldown, list, pulldownHandle } = this.classNames;
		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div style={outerDivStyle} className={root}>
					<div ref={infScroll => (this._infScroll = infScroll)} style={style}>
						{this.props.pullDownToRefresh && (
							<div className={pulldown} ref={pullDown => (this._pullDown = pullDown)}>
								<div
									className={pulldownHandle}
									style={{ top: -1 * this.maxPullDownDistance }}>
									{!this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="pullDownToRefresh" />
										</h3>
									)}
									{this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="releaseToRefresh" />
										</h3>
									)}
								</div>
							</div>
						)}
						<ul className={list}>{this.props.children}</ul>
						{!this.state.showLoader && !hasChildren && this.props.hasMore && <Spinner />}
						{this.state.showLoader && this.props.hasMore && <Spinner />}
						{!this.props.hasMore && this.props.endMessage}
					</div>
				</div>
			</IntlProvider>
		);
	}
}

export const InfiniteList = InfiniteListComponent;

InfiniteList.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

InfiniteListComponent.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string.isRequired,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

export default styled(customizable('InfiniteList', ['theme'])(InfiniteList), styles);
