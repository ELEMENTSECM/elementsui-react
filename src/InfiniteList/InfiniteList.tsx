import * as React from "react";
import { throttle, delay } from "lodash";
import BarLoader from "react-spinners/BarLoader";
import styled from "styled-components";
import InfiniteListItem from "./InfiniteListItem";
import { HotKeys } from "react-hotkeys";
import { keyMap, handlers } from "./keyboardNavigation";

export interface InfiniteListProps {
	/** DOM element's id attribute */
	id?: string;
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1; 2; 3] and then next load of data should be [1; 2; 3; 4; 5; 6]. */
	next?: () => void;
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore?: boolean;
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage?: React.ReactNode;
	/** Optional; give only if you want to have a fixed height scrolling content */
	height?: string | number;
	/** Optional; reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget?: React.ReactNode;
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not; if your children is not an array; specify this prop to tell if your items are 0 or more. */
	hasChildren?: boolean;
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled; so you may not receive as many events as you would expect. */
	onScroll?: (event?: any) => void;
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: number;
	/** Root div class name */
	className?: string;
	/** List class name */
	listClassName?: string;
	/** Load button class name */
	loadButtonClassName?: string;
	/** Spinner color in a form accepted in CSS stylesheet */
	spinnerColor?: string;
	/** HTML element tag that will rendered for this item. Default: <li> */
	tag?: keyof JSX.IntrinsicElements;
	/** Enable keyboard navigation. False by default. */
	keyboardNavigationEnabled?: boolean;
	/** Number of items to skip when navigating  */
	keyboardNavigationPageSize?: number;
}

type State = {
	dataLength: number;
	showLoader: boolean;
	actionTriggered: boolean;
	canScroll: boolean;
};

const Root = styled.div`
	overflow: hidden;
`; // Establish new Block Formatting Context

const List = styled.ul`
	box-sizing: border-box;
	list-style: none;
	margin: 0;
	padding: 0;
`;

const InfiniteListContainer = styled.div`
	box-sizing: border-box;
	overflow: auto;
	height: calc(100% - 4px);
	-webkit-overflow-scrolling: touch;
`;

const LoaderContainer = styled.div`
	box-sizing: border-box;
	overflow: hidden;
	height: 4px;
	display: flex;
`;

const LoadButton = styled.button`
	box-sizing: border-box;
	border: 0;
	padding: 0;
	width: 100%;
	background: #2180c0;
	> .caret {
		color: #fff;
	}
`;

export default class InfiniteList extends React.Component<InfiniteListProps, State> {
	static Item = InfiniteListItem;

	private el;
	private _infScroll;
	private _scrollableNode;
	private listRef = React.createRef<HTMLUListElement>();
	private keyboardHandlers;

	static defaultProps: Partial<InfiniteListProps> = {
		spinnerColor: "#2180c0",
		keyboardNavigationEnabled: false,
		keyboardNavigationPageSize: 3,
	};

	constructor(props: InfiniteListProps) {
		super(props);
		this.state = {
			dataLength: props.dataLength,
			showLoader: false,
			actionTriggered: false,
			canScroll: false,
		};

		this.keyboardHandlers = props.keyboardNavigationEnabled
			? handlers(this.listRef, { pageSize: this.props.keyboardNavigationPageSize })
			: null;
	}

	getScrollableTarget = () => {
		if (this.props.scrollableTarget instanceof HTMLElement) return this.props.scrollableTarget;
		if (typeof this.props.scrollableTarget === "string") {
			return document.getElementById(this.props.scrollableTarget);
		}
		if (this.props.scrollableTarget === null) {
			console.warn(
				`You are trying to pass scrollableTarget but it is null. This might happen because the element may not have been added to DOM yet.`
			);
		}
		return null;
	};

	private updateCanScroll = () => {
		const canScroll =
			!(this.el && this.listRef.current) || this.el.offsetHeight < this.listRef.current.offsetHeight;
		if (this.state.canScroll !== canScroll) {
			this.setState({ canScroll });
		}

		delay(this.updateCanScroll, 50);
	};

	private next = () => {
		this.setState((ps) => ({ ...ps, actionTriggered: true, showLoader: true }));
		this.props.next?.();
	};

	isElementAtBottom(target: HTMLElement) {
		const clientHeight =
			target === document.body || target === document.documentElement
				? window.screen.availHeight
				: target.clientHeight;

		return target.scrollHeight - Math.ceil(target.scrollTop) <= clientHeight;
	}

	onScrollListener = (event) => {
		setTimeout(() => this.props.onScroll && this.props.onScroll(event), 0);

		const target =
			this.props.height || this._scrollableNode
				? event.target as HTMLElement
				: document.documentElement.scrollTop
				? document.documentElement
				: document.body;

		if (this.state.actionTriggered) return;

		const atBottom = this.isElementAtBottom(target);

		if (atBottom && this.props.hasMore && this.props.dataLength) {
			this.next();
		}
	};

	private throttledOnScrollListener = throttle(this.onScrollListener, 150);

	static getDerivedStateFromProps(props, state) {
		if (props.dataLength === state.dataLength) return null;

		return {
			dataLength: props.dataLength,
			showLoader: false,
			actionTriggered: false,
		};
	}

	componentDidMount() {
		this._scrollableNode = this.getScrollableTarget();
		this.el = this.props.height ? this._infScroll : this._scrollableNode || window;
		this.el.addEventListener("scroll", this.throttledOnScrollListener);
		this.updateCanScroll();
	}

	componentWillUnmount() {
		this.el.removeEventListener("scroll", this.throttledOnScrollListener);
	}

	render() {
		const style = {
			height: this.props.height || "auto",
		};

		const hasChildren =
			this.props.hasChildren ||
			!!(this.props.children && (this.props.children as React.ReactNodeArray).length);

		const loading =
			(!this.state.showLoader && !hasChildren && this.props.hasMore) ||
			(this.state.showLoader && this.props.hasMore);

		return (
			<Root style={style}>
				<LoaderContainer>
					<BarLoader width="100%" height={4} color={this.props.spinnerColor} loading={loading} />
				</LoaderContainer>
				<InfiniteListContainer
					className={this.props.className}
					ref={(infScroll) => (this._infScroll = infScroll)}
				>
					<HotKeys keyMap={keyMap} handlers={this.keyboardHandlers}>
						<List
							ref={this.listRef}
							as={this.props.tag as any}
							className={this.props.listClassName}
							role="list"
						>
							{this.props.children}
						</List>
					</HotKeys>

					{!loading && this.props.hasMore && !this.state.canScroll && (
						<LoadButton onClick={this.next} className={this.props.loadButtonClassName}>
							<div className="caret" />
						</LoadButton>
					)}
					{!this.props.hasMore && this.props.endMessage}
				</InfiniteListContainer>
			</Root>
		);
	}
}
