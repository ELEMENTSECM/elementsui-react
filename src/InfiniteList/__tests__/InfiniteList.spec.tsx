import * as React from "react";
import { shallow } from "enzyme";
import InfiniteList from "..";

describe("InfiniteList", () => {
	test("displays correctly", () => {
		const wrapper = shallow(
			<InfiniteList dataLength={20} hasMore={false}>
				{new Array(20).fill(null).map((_, i) => {
					return <InfiniteList.Item key={i}>Item {i}</InfiniteList.Item>;
				})}
			</InfiniteList>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
