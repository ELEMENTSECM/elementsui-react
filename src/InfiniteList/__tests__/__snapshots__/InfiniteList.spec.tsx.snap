// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteList displays correctly 1`] = `
<styled.div
  style={
    Object {
      "height": "auto",
    }
  }
>
  <styled.div>
    <Loader
      color="#2180c0"
      css=""
      height={4}
      loading={false}
      speedMultiplier={1}
      width="100%"
    />
  </styled.div>
  <styled.div>
    <HotKeysEnabled
      handlers={null}
      keyMap={
        Object {
          "bottom": "end",
          "down": "down",
          "left": "left",
          "pageDown": "pagedown",
          "pageUp": "pageup",
          "right": "right",
          "top": "home",
          "up": "up",
        }
      }
    >
      <styled.ul
        role="list"
      >
        <Memo(InfiniteListItem)
          key="0"
        >
          Item 
          0
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="1"
        >
          Item 
          1
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="2"
        >
          Item 
          2
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="3"
        >
          Item 
          3
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="4"
        >
          Item 
          4
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="5"
        >
          Item 
          5
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="6"
        >
          Item 
          6
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="7"
        >
          Item 
          7
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="8"
        >
          Item 
          8
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="9"
        >
          Item 
          9
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="10"
        >
          Item 
          10
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="11"
        >
          Item 
          11
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="12"
        >
          Item 
          12
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="13"
        >
          Item 
          13
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="14"
        >
          Item 
          14
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="15"
        >
          Item 
          15
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="16"
        >
          Item 
          16
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="17"
        >
          Item 
          17
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="18"
        >
          Item 
          18
        </Memo(InfiniteListItem)>
        <Memo(InfiniteListItem)
          key="19"
        >
          Item 
          19
        </Memo(InfiniteListItem)>
      </styled.ul>
    </HotKeysEnabled>
  </styled.div>
</styled.div>
`;
